### PR TITLE
Add integration tests

### PR DESCRIPTION
**Pull request by mrcnski**

### CODE REVIEW

File: tests/integration_test.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+	skynet "github.com/NebulousLabs/go-skynet"
+	"gopkg.in/h2non/gock.v1"
+)
+
+// TestUploadAndDownloadFile tests uploading and downloading a single file.
+func TestUploadAndDownloadFile(t *testing.T) {
+	defer gock.Off() // Flush pending mocks after test execution
+
+	const srcFile = "../testdata/file1.txt"
```
Comment (MSevey) [inline tests/integration_test.go:18]: Looks like this code is already duplicated between these two test and I imagine it will be used for any future testing. Can we extract it into a helper similar to the `siatest TestGroups`?

**Comment by MSevey** [comment]
LGTM, if @peterjanbrone agrees.

We should add some merge approval requirements to github probably

File: tests/integration_test.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	"path"
+	"path/filepath"
+	"testing"
+
+	skynet "github.com/NebulousLabs/go-skynet"
+	"gopkg.in/h2non/gock.v1"
+)
+
+// TestUploadAndDownloadFile tests uploading and downloading a single file.
+func TestUploadAndDownloadFile(t *testing.T) {
```
Comment (peterjan) [inline tests/integration_test.go:31]: I'm not sure how these tests validate the inner workings of the library logic itself. Perhaps I don't fully understand but it looks to me it only verifies the most basic case and ignores the data that's being sent, request headers, query string params etc?

For example the `srcFile`, does that file exist? Are we actually mocking the post body? I think `gock` is capable of all of this.

File: tests/integration_test.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	gock.New(skynet.DefaultUploadOptions.PortalURL).
+		Post(skynet.DefaultUploadOptions.PortalUploadPath).
+		Reply(200).
+		JSON(map[string]string{"skylink": "testskynet"})
+
+	skylink2, err := skynet.UploadFile(srcFile, skynet.DefaultUploadOptions)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if skylink2 != skylink {
```
Comment (peterjan) [inline tests/integration_test.go]: Does this test pass? From reading the code I would expect it to fail because the mocked response is `{"skylink": "testskynet"}` and not `{"skylink": "sia://testskynet"}` which is what you are asserting, right?

File: tests/integration_test.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+		Reply(200).
+		BodyString("test\n")
+
+	file, err := ioutil.TempFile("", t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	dstFile := file.Name()
+
+	err = skynet.DownloadFile(dstFile, skylink, skynet.DefaultDownloadOptions)
```
Comment (peterjan) [inline tests/integration_test.go]: This should verify the behaviour of `DownloadFile`. For example here it should verify the prefix is being trimmed from the given skylink when the request is being formed.

```
	resp, err := http.Get(fmt.Sprintf("%s/%s", strings.TrimRight(opts.PortalURL, "/"), strings.TrimPrefix(skylink, "sia://")))
	if err != nil {
```

we should probably also try and mock error responses and cover how the sdk handles those

File: tests/integration_test.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	// Check file equality.
+	f1, err1 := ioutil.ReadFile(srcFile)
+	if err1 != nil {
+		t.Fatal(err1)
+	}
+	f2, err2 := ioutil.ReadFile(path.Clean(dstFile))
+	if err2 != nil {
+		t.Fatal(err2)
+	}
+	if !bytes.Equal(f1, f2) {
```
Comment (peterjanbrone) [inline tests/integration_test.go:63]: Is `gock` actually writing data to our `dstfile`? Where do we instruct it to do that? That's pretty neat but I don't understand where all of this is being configured

File: tests/integration_test.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		t.Fatalf("Downloaded file at %v did not equal uploaded file %v", dstFile, srcFile)
+	}
+}
+
+// TestUploadDirectory tests uploading an entire directory.
+func TestUploadDirectory(t *testing.T) {
+	defer gock.Off()
+
+	const srcDir = "../testdata"
```
Comment (peterjanbrone) [inline tests/integration_test.go:72]: could it be you've forgotten to commit this `testdata` folder because it's  in the gitignore or something?

**Review by peterjan — CHANGES_REQUESTED** [review]

File: tests/integration_test.go (hunk 5 above)
Comment (peterjan) [inline tests/integration_test.go:103]: This is pretty neat but really `file1.txt` just contains the text "test". I wonder if `gock` is capable of doing more advanced stuff. I think we should look into it a bit more. I will do that as well to see what we could do to improve our testing coverage.

File: tests/integration_test.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	if err2 != nil {
+		t.Fatal(err2)
+	}
+	if !bytes.Equal(f1, f2) {
+		t.Fatalf("Downloaded file at %v did not equal uploaded file %v", dstFile, srcFile)
+	}
+}
+
+// TestUploadDirectory tests uploading an entire directory.
+func TestUploadDirectory(t *testing.T) {
```
Comment (peterjan) [inline tests/integration_test.go:114]: Nice that we have this now but I don't feel it really covers any of the logic within `UploadDirectory`. We should structure the test in order to verify things like `walkDirectory` or the way the formdata is being posted to the backend.

**Comment by peterjan** [comment]
I think this is a great start. I feel that most of the test cases are missing coverage around the edge cases. Overall it feels like we are not truly testing any of the logic inside those SDK functions. These tests are really meant to provide us with peace of mind when somebody alters the `UploadDirectory` function. If for example I were to break the `walkDirectory` logic, this test should also break. I think it would not do that in its current form.

File: tests/integration_test.go (hunk 3 above)
Comment (mrcnski) [reply tests/integration_test.go]: It passes because `UploadFile` prepends the sia prefix to the response.

Comment (peterjan) [reply tests/integration_test.go]: Oh ok, well then this is a great test actually ^^

File: testdata/file2.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+foo2
```
Comment (peterjan) [inline testdata/file2.txt:1]: Are the contents of these files important at all? Should/could we just write tmp files to disk in the test setup and then remove them in the teardown?

**Review by peterjan — APPROVED** [review]

File: tests/integration_test.go [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+		bytes, _ := httputil.DumpRequestOut(request, true)
+		interceptedRequest = string(bytes)
+	}
+
+	// interceptedRequest contains the raw data of intercepted requests.
+	interceptedRequest string
+)
+
+// TestUploadAndDownloadFile tests uploading and downloading a single file.
+func TestUploadAndDownloadFile(t *testing.T) {
```
Comment (peterjan) [inline tests/integration_test.go:31]: I would split this up

File: testdata/file2.txt (hunk 8 above)
Comment (mrcnski) [reply testdata/file2.txt:1]: I think this is simpler than setting up the test directory structure before every test. It's especially important for testing walkDirectory
